flink-libraries/flink-table/src/test/scala/org/apache/flink/table/plan/rules/util/RexProgramProjectExtractorTest.scala

```diff
@@ -55,6 +56,34 @@ class RexProgramProjectExtractorTest {
   }
 
   @Test
+  def testExtractRefNestedInputFields(): Unit = {
+    val rexProgram = buildRexProgramWithNesting()
+    val usedFields = extractRefInputFields(rexProgram)
+    val usedNestedFields = extractRefNestedInputFields(rexProgram, usedFields)
+    val expected = Array[Array[String]](Array("amount"), Array("*"))
+    assertThat(usedNestedFields, is(expected))
+  }
+
+  @Test
+  def testExtractRefNestedInputFieldsWithNoNesting(): Unit = {
+    val rexProgram = buildRexProgram()
+    val usedFields = extractRefInputFields(rexProgram)
+    val usedNestedFields = extractRefNestedInputFields(rexProgram, usedFields)
+    val expected = Array[Array[String]](Array("*"), Array("*"), Array("*"))
+    assertThat(usedNestedFields, is(expected))
+  }
+
+  @Test
+  def testExtractDeepRefNestedInputFields(): Unit = {
+    val rexProgram = buildRexProgramWithDeepNesting()
+    val usedFields = extractRefInputFields(rexProgram)
+    val usedNestedFields = extractRefNestedInputFields(rexProgram, usedFields)
+    val expected = Array[Array[String]](Array("amount"), Array("passport.status"))
```

Comment:
It would be good to have a test where the Array with the nested fields contains more than one entry.